5.0/en/0x12-V3-Session-management.md

```diff
@@ -91,10 +90,88 @@ Attackers start a half-open attack by attempting to lock, reset, or recover a cr
 | :---: | :--- | :---: | :---: | :---: | :---: | :---: |
 | **3.7.1** | [MODIFIED] Verify the application requires re-authentication or secondary factor verification before allowing any sensitive transactions or account modifications. | ✓ | ✓ | ✓ | 306 | |
 
+
+
+## V3.8 OAuth 2.0 Protocol
+
+This section describes the best current security practices for OAuth 2.0 as derived from its RFC. OAuth became the standard for API protection and the basis for federated login using OpenID Connect. OpenID Connect 1.0 is a simple identity layer on top of the OAuth 2.0 protocol. It enables clients to verify the identity of the end-user based on the authentication performed by an authorization server, as well as to obtain basic profile information about the end-user in an interoperable and REST-like manner. 
+
+## Terminology: 
+
+* `Access tokens` --- provides an abstraction, replacing different authorization constructs (e.g., username and password, assertion) for a single token understood by the resource server. This abstraction enables issuing access tokens valid for a short time period, as well as removing the resource server's need to understand a wide range of authentication schemes. 
+* `Refresh tokens` --- are credentials used to obtain access tokens. These are issued to the client by the authorization server and are used to obtain a new access token when the current access token becomes invalid or expires, or to obtain additional access tokens with identical or narrower scope (access tokens may have a shorter lifetime and fewer permissions than authorized by the resource owner). 
+* `Client` --- generally refers to an application making protected resource requests on behalf of the resource owner and with its authorization. The term "client" does not imply any particular implementation characteristics (e.g., whether the application executes on a server, a desktop, or other devices). 
+* `Authorization Server (AS)` --- refers to the server issuing access tokens to the client after successfully authenticating the resource owner and obtaining authorization.
+* `Resource Owner (RO)` --- refers to an entity capable of granting access to a protected resource. When the resource owner is a person, it is referred to as an end-user. 
+* `Resource Server (RS)` --- refers to the server hosting the protected resources, capable of accepting and responding to protected resource requests using access tokens. 
+
+
```

Comment:
Do you want something like a column or a marker that says who each specific requirement is for such as the developer of the Client, the developer of the Authorization Server, the developer of the Resource Server, etc?

Reply:
Ideally I would like each to be in a different section like 3.8.x, 3.9.x, 3.10.x, etc.

Reply:
Hmmm, let me see if I can devise that in a way to fit that preference. I grouped them according to the "topics"/OAuth use-cases.

Reply:
@tghosth, here is an example of what I think can work as a header to distinguish the main implementor concerned for each of the new OAuth 2.0 requirements. What do you think?

> 
| # | Description | L1 | L2 | L3 | Implementor |
| :---: | :--- | :---: | :---: | :---: | :---: |  
| **3.8.1** | [ADDED] Verify that the Authorization Server supports CSRF protection either using the mechanism provided by PKCE, the "nonce" parameter or the "state" parameter. | ✓ | ✓ | ✓ | Authorization Server Developer |

Reply:
I don't really want to add another column, I think I would prefer a section title.

Reply:
Ok, let me work some magic on it by grouping things together.

Reply:
Hi @tghosth, 

I'm having a hard time making your request for making it a per implementor-based list of requirements given the subsections of the proposed OAuth-related requirements. Another problem I have seen is there would be duplicate requirements if we will do it on an implementor-based list only such as in the following examples: 


## OAuth 2.0 Essential Basics

Some important Info text here. 

**Authorization Server Developer** 
| # | Description | L1 | L2 | L3 |
| :---: | :--- | :---: | :---: | :---: |
| **3.8.1** | [ADDED] Verify that the Authorization Server supports CSRF protection either using the mechanism provided by PKCE, the "nonce" parameter or the "state" parameter. | ✓ | ✓ | ✓ | 
| **3.8.4** | [ADDED] Verify that if a Client sends a valid PKCE "code_challenge" parameter in the authorization request, the Authorization Server enforces the correct usage of "code_verifier" at the token endpoint. | ✓ | ✓ | ✓ |
| **3.8.5** | [ADDED] Verify that Authorization Servers are mitigating PKCE Downgrade Attacks by ensuring a token request containing a "code_verifier" parameter is accepted only if a "code_challenge" parameter was present in the authorization request. | ✓ | ✓ | ✓ |


**Client Developer**
| # | Description | L1 | L2 | L3 |
| :---: | :--- | :---: | :---: | :---: |
| **3.8.2** | [ADDED] Verify that when an OAuth Client can interact with more than one Authorization Server, Clients should verify that the issuer "iss" parameter value is what it expected from the authorization response to prevent against mix-up attacks. [reference link for Mix-up attacks: https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics-18#mix_up] In the absence of "iss" parameter, Clients may instead use distinct redirect URIs to identify authorization endpoints and token endpoints.  | ✓ | ✓ | ✓ | 


## PKCE - Proof Key for Code Exchange Mechanism / Authorization Code Grant

Info text here about PKCE. 

**Authorization Server Developer** 
| # | Description | L1 | L2 | L3 |
| :---: | :--- | :---: | :---: | :---: |
| **3.8.3** | [ADDED] Verify that the replay of authorization codes into the authorization response is prevented either by using the PKCE flow or alternatively the OpenID Connect "nonce" parameter and the respective Claim in the ID Token. The PKCE challenge or OpenID Connect "nonce" must be transaction-specific and securely bound to the client and the user agent in which the transaction was started.  | ✓ | ✓ | ✓ |

**Client Developer**
| # | Description | L1 | L2 | L3 |
| :---: | :--- | :---: | :---: | :---: |
| **3.8.3** | [ADDED] Verify that the replay of authorization codes into the authorization response is prevented either by using the PKCE flow or alternatively the OpenID Connect "nonce" parameter and the respective Claim in the ID Token. The PKCE challenge or OpenID Connect "nonce" must be transaction-specific and securely bound to the client and the user agent in which the transaction was started.  | ✓ | ✓ | ✓ |


Requirement **3.8.3** has to be implemented both by the authorization server developer and the client developer. It is going to be duplicated if we do it this way as presented if we will do them by implementor. 

Also, notice that while **3.8.1** fits in the OAuth Essentials subsection, **3.8.4** and **3.8.5** falls under PKCE making it a little bit confusing to read while it is grouped well under Authorization Server Developer as the main implementor of those controls. 

Another solution I can think of is have all the subsections explained on top such as PKCE, Basics, Token Replay Prevention, Access Token Privilege Restriction, etc. and then after all of those texts were stated with their relevant info we list all the implementor-based controls list such as shown here, but it still doesn't solve the duplicates though. 

I would propose going back to having a suffixed "Implementor" column for each requirement to avoid duplicates and misplaced requirements/groupings. Let me know your thoughts.

Reply:
### 3.8.3

If the Authorization Server (AS) implements this functionality, does the client have any choice whether to implement it or not? It feels like this is just a requirement on the AS.

Even if the client does have the choice whether to implement it or not, the requirement on the client is different because the client just needs to make sure it is using the functionality, not create the functionality.

### 3.8.4/3.8.5

I'm not sure I see the difference between these requirements

### 3.8.1

I'm not sure I understand the question. It seems like an AS requirement.


---

Overall, I really don't want to have another column here and I think we need to try and make this "who is the implementor" framework work.

Reply:
Alright, I think I have a bit of an idea how to make it a little bit cleaner even if it seems messy. Stay tuned for a new commit next week for this PR!